README.md
# LotteryPlatinum
一个从奇葩角度分析彩票的工具

## 使用说明

执行程序后，有3个选项。说明如下：

`1`  给一注双色球号码，搜索该号码在所有历史记录中中过的最高奖项。**判断依据为总匹配号码数量。所以当二等奖和三等奖同时存在时，会同时展现。其他情况会显示最高奖项**

 ---
 
 `2` 统计最近n期中奖号码中，每个号码出现的次数。输入数字以确定期数，不输入的话直接回车默认为100期。
 
 ---
 
 `3` 更新当前数据库，**需要提前准备聚合数据的app key**，后期可能会通过代码实现定期更新数据库。有key的可以自行更新，没有的话，请暂时以git的数据库为准。
 
 


data/history.txt
[{"no": "20091", "red": ["01", "09", "11", "12", "16", "19"], "blue": "16"}, {"no": "20090", "red": ["02", "21", "23", "26", "31", "32"], "blue": "06"}, {"no": "20089", "red": ["02", "14", "16", "21", "29", "30"], "blue": "10"}, {"no": "20088", "red": ["01", "06", "12", "26", "29", "30"], "blue": "12"}, {"no": "20087", "red": ["11", "15", "20", "23", "25", "33"], "blue": "10"}, {"no": "20086", "red": ["02", "04", "06", "15", "24", "27"], "blue": "06"}, {"no": "20085", "red": ["01", "02", "05", "09", "19", "24"], "blue": "16"}, {"no": "20084", "red": ["03", "07", "16", "17", "23", "30"], "blue": "07"}, {"no": "20083", "red": ["01", "19", "25", "26", "30", "31"], "blue": "12"}, {"no": "20082", "red": ["02", "08", "11", "17", "21", "30"], "blue": "09"}, {"no": "20081", "red": ["01", "05", "13", "14", "27", "33"], "blue": "15"}, {"no": "20080", "red": ["14", "15", "18", "22", "31", "33"], "blue": "01"}, {"no": "20079", "red": ["05", "12", "20", "21", "22", "29"], "blue": "14"}, {"no": "20078", "red": ["03", "11", "14", "16", "21", "32"], "blue": "04"}, {"no": "20077", "red": ["03", "10", "16", "21", "25", "27"], "blue": "12"}, {"no": "20076", "red": ["10", "15", "16", "18", "20", "27"], "blue": "06"}, {"no": "20075", "red": ["03", "11", "13", "20", "24", "30"], "blue": "16"}, {"no": "20074", "red": ["04", "08", "09", "13", "19", "33"], "blue": "12"}, {"no": "20073", "red": ["05", "07", "11", "13", "27", "29"], "blue": "03"}, {"no": "20072", "red": ["06", "08", "10", "15", "17", "26"], "blue": "04"}, {"no": "20071", "red": ["09", "11", "12", "13", "22", "23"], "blue": "08"}, {"no": "20070", "red": ["01", "02", "04", "06", "19", "21"], "blue": "15"}, {"no": "20069", "red": ["03", "09", "10", "13", "18", "26"], "blue": "04"}, {"no": "20068", "red": ["12", "16", "21", "26", "27", "32"], "blue": "10"}, {"no": "20067", "red": ["04", "07", "09", "23", "27", "30"], "blue": "08"}, {"no": "20066", "red": ["02", "09", "13", "17", "26", "28"], "blue": "07"}, {"no": "20065", "red": ["09", "15", "18", "21", "23", "26"], "blue": "08"}, {"no": "20064", "red": ["01", "03", "07", "21", "27", "32"], "blue": "01"}, {"no": "20063", "red": ["12", "15", "16", "22", "29", "32"], "blue": "14"}, {"no": "20062", "red": ["10", "14", "17", "22", "26", "27"], "blue": "05"}, {"no": "20061", "red": ["08", "17", "24", "26", "27", "31"], "blue": "04"}, {"no": "20060", "red": ["05", "09", "14", "20", "24", "30"], "blue": "08"}, {"no": "20059", "red": ["02", "04", "10", "17", "22", "25"], "blue": "14"}, {"no": "20058", "red": ["01", "03", "11", "12", "19", "26"], "blue": "07"}, {"no": "20057", "red": ["09", "14", "21", "23", "26", "32"], "blue": "03"}, {"no": "20056", "red": ["02", "05", "08", "12", "26", "31"], "blue": "14"}, {"no": "20055", "red": ["01", "05", "07", "23", "28", "30"], "blue": "12"}, {"no": "20054", "red": ["03", "10", "19", "25", "26", "31"], "blue": "02"}, {"no": "20053", "red": ["02", "14", "15", "16", "32", "33"], "blue": "01"}, {"no": "20052", "red": ["02", "08", "13", "29", "32", "33"], "blue": "15"}, {"no": "20051", "red": ["03", "06", "08", "11", "19", "28"], "blue": "08"}, {"no": "20050", "red": ["04", "09", "17", "20", "32", "33"], "blue": "15"}, {"no": "20049", "red": ["01", "07", "09", "12", "18", "22"], "blue": "04"}, {"no": "20048", "red": ["12", "14", "18", "23", "30", "32"], "blue": "02"}, {"no": "20047", "red": ["04", "10", "17", "19", "28", "32"], "blue": "01"}, {"no": "20046", "red": ["13", "19", "25", "26", "27", "32"], "blue": "08"}, {"no": "20045", "red": ["02", "03", "15", "21", "22", "33"], "blue": "16"}, {"no": "20044", "red": ["07", "08", "16", "17", "19", "24"], "blue": "07"}, {"no": "20043", "red": ["01", "11", "24", "25", "27", "30"], "blue": "07"}, {"no": "20042", "red": ["02", "06", "07", "11", "14", "31"], "blue": "03"}, {"no": "20041", "red": ["06", "07", "16", "22", "24", "25"], "blue": "13"}, {"no": "20040", "red": ["01", "06", "12", "16", "19", "21"], "blue": "04"}, {"no": "20039", "red": ["02", "09", "10", "11", "16", "29"], "blue": "02"}, {"no": "20038", "red": ["01", "06", "07", "18", "23", "24"], "blue": "15"}, {"no": "20037", "red": ["01", "04", "11", "13", "17", "24"], "blue": "15"}, {"no": "20036", "red": ["07", "09", "16", "22", "24", "32"], "blue": "06"}, {"no": "20035", "red": ["10", "14", "24", "25", "28", "33"], "blue": "11"}, {"no": "20034", "red": ["02", "08", "15", "16", "26", "32"], "blue": "03"}, {"no": "20033", "red": ["07", "10", "12", "21", "31", "32"], "blue": "01"}, {"no": "20032", "red": ["03", "11", "13", "14", "15", "26"], "blue": "13"}, {"no": "20031", "red": ["02", "05", "09", "15", "16", "27"], "blue": "09"}, {"no": "20030", "red": ["17", "18", "21", "29", "30", "32"], "blue": "03"}, {"no": "20029", "red": ["01", "12", "18", "20", "30", "32"], "blue": "05"}, {"no": "20028", "red": ["05", "06", "15", "18", "26", "32"], "blue": "08"}, {"no": "20027", "red": ["12", "13", "14", "24", "25", "28"], "blue": "06"}, {"no": "20026", "red": ["01", "04", "09", "18", "19", "31"], "blue": "06"}, {"no": "20025", "red": ["01", "09", "10", "15", "20", "29"], "blue": "02"}, {"no": "20024", "red": ["10", "13", "27", "28", "29", "31"], "blue": "08"}, {"no": "20023", "red": ["04", "05", "20", "21", "30", "33"], "blue": "08"}, {"no": "20022", "red": ["02", "10", "20", "22", "25", "30"], "blue": "02"}, {"no": "20021", "red": ["02", "06", "14", "16", "27", "31"], "blue": "07"}, {"no": "20020", "red": ["01", "06", "09", "13", "15", "22"], "blue": "14"}, {"no": "20019", "red": ["15", "19", "27", "28", "30", "33"], "blue": "03"}, {"no": "20018", "red": ["05", "07", "08", "11", "17", "22"], "blue": "13"}, {"no": "20017", "red": ["02", "04", "07", "15", "20", "27"], "blue": "04"}, {"no": "20016", "red": ["05", "06", "08", "17", "24", "27"], "blue": "07"}, {"no": "20015", "red": ["08", "09", "22", "24", "30", "33"], "blue": "01"}, {"no": "20014", "red": ["02", "07", "08", "10", "12", "16"], "blue": "07"}, {"no": "20013", "red": ["02", "08", "10", "20", "21", "30"], "blue": "1

1000w.py
# coding=utf-8
import requests, json
from os import system
system("title 白金之星-另类分析工具")

# 白金之星-双色球分析工具-v1.0
# author Hermit_xx
# Email: medivh_yx@qq.com

# 此处输入聚合数据的app_key
app_key = ''


def save(s):
    f = open('data\\history.txt', 'w')
    f.write(s)
    f.close()


def update(key):
    url = 'http://apis.juhe.cn/lottery/history'
    para = {'key': app_key,
            'lottery_id': 'ssq',
            'page_size': '50',
            }
    all_list = []
    # 取第一页
    r = requests.get(url, params=para)
    # print('HTTP状态码:', r.status_code)
    # print('返回内容:',r.text)
    res = json.loads(r.text)
    if (str(res['error_code']) != '0'):
        print('出现异常，错误码：', res['error_code'])
        return False
    print('总页数:', res['result']['totalPage'])
    total_page = int(res['result']['totalPage']) + 1

    # 遍历内容
    for item in res['result']['lotteryResList']:
        number = item['lottery_res'].split(',')
        save_item = {'no': item['lottery_no'],
                     'red': number[0:6],
                     'blue': number[-1]
                     }
        all_list.append(save_item)

    f = open('data\\part\\1.txt', 'w')
    f.write(json.dumps(all_list))
    f.close()

    # 遍历2+页
    for index in range(2, total_page):
        part_list = []
        print('正在爬取第', str(index), '页')
        para['page'] = index
        r = requests.get(url, params=para)
        res = json.loads(r.text)
        if (str(res['error_code']) != '0'):
            print('出现异常，错误码：', res['error_code'])
            return False

        res = json.loads(r.text)

        # 遍历内容
        for item in res['result']['lotteryResList']:
            number = item['lottery_res'].split(',')
            save_item = {'no': item['lottery_no'],
                         'red': number[0:6],
                         'blue': number[-1]
                         }
            part_list.append(save_item)
            all_list.append(save_item)

        f = open('data\\part\\' + str(index) + '.txt', 'w')
        f.write(json.dumps(part_list))
        f.close()

    save(json.dumps(all_list))
    print('更新完成！')


def moneyConvert(s):
    if s == '6+1':
        return '一等奖'
    elif s == '6+0':
        return '二等奖'
    elif s == '5+1':
        return '三等奖'
    elif s in ('5+0', '4+1'):
        return '四等奖'
    elif s in ('4+0', '3+1'):
        return '五等奖'


def find():

    no = input('输入号码，先输入红球，红球大小无需排序，最后输入蓝球，每个数字之间用空格分隔。\n输入举例 01 02 03 04 15 19 16\n')
    arr = no.split(' ')
    if (len(arr[:-1]) != 6):
        print('输入有误，重新输入。\n')
        return False
    else:
        print('红球：', arr[:-1], '蓝球：', arr[-1])

    custom_red = arr[:-1]
    custom_blue = arr[-1]

    global data_arr

    max_count = 0

    max_list = []

    # 利用三重积分、曲面积分、离散傅里叶级数公式构造核心算法
    for item in data_arr:
        red = len(set(custom_red).intersection(set(item['red'])))
        if custom_blue == item['blue']:
            blue = 1
        else:
            blue = 0
        if red + blue <= 3:
            continue
        count = red + blue
        if count > max_count:
            max_count = count
            item['get'] = str(red) + '+' + str(blue)
            max_list = [item, ]
        elif count == max_count:
            item['get'] = str(red) + '+' + str(blue)
            max_list.append(item)

    if max_count == 0:
        print('该注号码在历史中奖记录中没有大于 3 个号码相符的情况。')
    else:
        print('该注号码在历史中奖记录中最多有', max_count, '个号码正确！')
        print('测试号码：' + ' '.join(custom_red) + '-' + custom_blue)
        print('最高奖中奖次数：', len(max_list))
        for item in max_list:
            print('期数：', item['no'], '中奖号码：', ' '.join(item['red'])
                  + '-' + item['blue'], '符合条件：', item['get'],
                  '奖级：', moneyConvert(item['get']))


def cal(limit=100):
    if(limit == ''):
        limit = 100
    else:
        limit = int(limit)
    f = open('data\\history.txt', 'r')
    json_str = f.read()
    data_arr = json.loads(json_str)
    red_stat = {}
    blue_stat = {}
    first_no = ''
    last_no = ''
    for i in range(1, 34):
        red_stat[str(i)] = 0
    for i in range(1, 17):
        blue_stat[str(i)] = 0
    index = 0
    for item in data_arr:
        if index == 0:
            first_no = item['no']
        for red_ball in item['red']:
            red_stat[str(int(red_ball))] += 1
        blue_stat[str(int(item['blue']))] += 1
        index += 1
        if index > limit:
            last_no = item['no']
            break
    print('开始期数：', first_no)
    print('结束期数：', last_no)

    red_stat = sorted(red_stat.items(), key=lambda item: item[1])
    blue_stat = sorted(blue_stat.items(), key=lambda item: item[1])
    # print('红球统计结果：', red_stat)
    # print('蓝球统计结果：', blue_stat)
    print('最近', limit, '期红球出现次数统计(由多到少排序)')
    for i in range(1, 34):
        print(red_stat[i * -1][0], ':出现', red_stat[i * -1][1], '次')

    print('最近', limit, '期蓝球出现次数统计(由多到少排序)')
    for i in range(1, 17):
        print(blue_stat[i * -1][0], ':出现', blue_stat[i * -1][1], '次')

def loadData():
    global data_arr
    f = open('data\\history.txt', 'r')
    json_str = f.read()
    data_arr = json.loads(json_str)
    print('当前数据库最新双色球期数为：', data_arr[1]['no'])

if __name__ == '__main__':

    loadData()

    while True:
        mode = input('======Link Start!======\n请输入功能序号\n1、辉煌历史：某注号码历史记录中最高奖\n2、硬核分析：分析最近n期每个号码出现的次数\n'
                     '3、更新数据（需要有聚合数据的app_key）\n')
        if(mode == '1'):
            find()
        elif(mode == '2'):
            mode = input('请输入期数范围，输入50表示统计最近50期号码。\n不输入默认为100期。')
            cal(mode)
        elif(mode == '3'):
            key = input(app_key)
            update(app_key)
            loadData()

